dask_planner/src/sql.rs

```diff
@@ -576,3 +571,112 @@ impl PlanVisitor for OptimizablePlanVisitor {
         Ok(true)
     }
 }
+
+fn generate_numeric_signatures(n: i32) -> Signature {
+    // Generates all combinations of vectors of length n,
+    // i.e., the Cartesian product
+    let datatypes = vec![
+        DataType::Int8,
+        DataType::Int16,
+        DataType::Int32,
+        DataType::Int64,
+        DataType::UInt8,
+        DataType::UInt16,
+        DataType::UInt32,
+        DataType::UInt64,
+        DataType::Float16,
+        DataType::Float32,
+        DataType::Float64,
+    ];
+    let mut cartesian_setup = vec![];
+    // cartesian_setup = [datatypes, datatypes] when n == 2, etc.
+    for _ in 0..n {
+        cartesian_setup.push(datatypes.clone());
+    }
+
+    let mut exact_vector = vec![];
+    let mut datatypes_iter = cartesian_setup.iter();
+    // First pass
+    if let Some(first_iter) = datatypes_iter.next() {
+        for datatype in first_iter {
+            exact_vector.push(vec![datatype.clone()]);
+        }
+    }
+    // Generate list of lists with length n
+    for iter in datatypes_iter {
+        let mut outer_temp = vec![];
+        for outer_datatype in exact_vector {
+            for inner_datatype in iter {
+                let mut inner_temp = outer_datatype.clone();
+                inner_temp.push(inner_datatype.clone());
+                outer_temp.push(inner_temp);
+            }
+        }
+        exact_vector = outer_temp;
+    }
+
+    // Create vector of TypeSignatures
+    let mut one_of_vector = vec![];
+    for vector in exact_vector.iter() {
+        one_of_vector.push(TypeSignature::Exact(vector.clone()));
+    }
+
+    Signature::one_of(one_of_vector.clone(), Volatility::Immutable)
+}
+
+#[allow(dead_code)]
+fn generate_signatures(cartesian_setup: Vec<Vec<DataType>>) -> Signature {
```

Comment:
I've marked this with `#[allow(dead_code)]` since it will be used with #813, #838, #876, which haven't been merged yet.

Reply:
Wonder if `generate_numeric_signatures` could be refactored to use this function to generate the list of numeric signatures?